stdlib/public/core/Hashing.swift

```diff
@@ -214,3 +194,121 @@ func _combineHashValues(_ firstValue: Int, _ secondValue: Int) -> Int {
   x ^= UInt(bitPattern: secondValue) &+ magic &+ (x &<< 6) &+ (x &>> 2)
   return Int(bitPattern: x)
 }
+
+/// An unsafe wrapper around a stateful hash function, presenting a faux purely
+/// functional interface to eliminate ARC overhead.
+///
+/// This is not a true value type; calling `appending` or `finalized` actually
+/// mutates `self`'s state.
+@_fixed_layout
+public struct _UnsafeHasher {
+  @_versioned
+  internal let _rawState: UnsafeMutableRawPointer
+
+  internal var _state: UnsafeMutablePointer<_Hasher> {
+    @inline(__always)
+    get { return _rawState.assumingMemoryBound(to: _Hasher.self) }
+  }
+
+  @inline(__always)
+  @_versioned
+  internal init(_ state: UnsafeMutablePointer<_Hasher>) {
+    self._rawState = UnsafeMutableRawPointer(state)
+  }
+
+  @_versioned
+  // not @_inlineable
+  @effects(readonly) // FIXME(hasher): Unjustified
+  static func hashValue<H: Hashable>(for pointer: UnsafePointer<H>) -> Int {
+    var hasher = _Hasher()
+    return withUnsafeMutablePointer(to: &hasher) { p in
+      return pointer.pointee._hash(into: _UnsafeHasher(p))._finalized()
+    }
+  }
+
+  @_versioned
+  // not @_inlineable
+  @effects(readonly)
+  internal func appending(bits value: UInt) -> _UnsafeHasher {
+    // The effects attribute is a lie; however, it enables the compiler to
+    // eliminate unnecessary retain/releases protecting Hashable state around
+    // calls to `_Hasher.append(_:)`.
+    //
```

Comment:
I had a whole bunch of thoughts earlier on this particular design, but I think everyone who's commented is in agreement so far about how we feel about it. Hopefully, the new `@effects` annotation will fall into place swiftly.

It's not merely for the sake of elegance, but I think the whole design of the API suffers in how comprehensible it is to the reader when what's clearly a stateful object being mutated (after all, the whole point of this PR is to enable a stateful hasher--or at least that's my understanding of it) is masquerading as a value type.

Reply:
The fun thing is that a fully immutable SipHash implementation (with a 100% valid `@effect(readonly)` declaration) actually performed surprisingly well, considering it had to keep passing around five words. In use it looked exactly the same as this one, except it actually was a real value type.

However, its performance relied on the size/layout of the hasher state being part of the ABI -- hence the clever replacement. Struct resiliency may cause performance problems for the `@effects(releasenone)` mutating case, too -- we'll have to find out.